### PR TITLE
feat(router): truncate routing tokens at configured media boundaries

### PR DESCRIPTION
**Pull request by slin1237**
### Problem
cache_aware matches the full wire `input_ids`, but on multimodal lanes an engine can only share prefix cache across conversations up to the first media placeholder (engine cache keys incorporate media content). Match ratios computed over the full sequence therefore shrink as conversations grow: on a production multimodal video lane, a conversation follow-up (~17k wire tokens containing its ~3k first-turn text) scores matched/raw ~= 0.17 and never returns to the engine holding its conversation state — measured at under 0.1% deep-prefix conversion for the ~37% of traffic that is follow-ups, while an automated fan-out cohort with near-identical prefixes (matched/raw >= 0.92) converts at 78%. No fixed threshold fixes this: third turns score lower still.
### Solution
`--routing-token-boundaries <ids>`: truncate extracted routing tokens at the first occurrence of any configured boundary id (e.g. media placeholder ids) before worker selection, on all three selection paths (HTTP regular, HTTP PD, gRPC pipeline). Matching then covers exactly the engine-shareable, conversation-invariant prefix: a follow-up scores ~1.0 against its own conversation at any length, unrelated requests score near zero. Empty config is byte-identical to today. A boundary in first position yields an empty prefix (match rate 0 → least-loaded), avoiding full-text fallback into the string tree. Truncations are observable via `smg_routing_tokens_truncated_total{router_type}`.

## Changes
- `routers/common/routing_tokens.rs`: shared truncation (borrowed-slice and in-place-owned variants) + metric
- `routers/http/router.rs`: truncation at both token-extraction sites before `SelectWorkerInfo`
- `routers/http/pd_router.rs`: `/generate` routing-input extraction factored into `generate_routing_inputs` with truncation
- gRPC: truncation at the `WorkerSelectionStage` choke point; boundaries plumbed through `PipelineDeps`
- Config: `routing_token_boundaries: Vec<u32>` (types/builder/CLI space-separated list, serde default) + flow test
- Python bindings: field appended at the positional tail (lib.rs x5, `router_args.py`, frozen-sequence tests)

## Composition
The obligation from #2202 is discharged in this PR: `route_streaming_request` truncates hinted tokens at the configured boundaries, and a boundary-emptied hint falls back to buffered routing (body unconsumed). No coordination needed with open PR #2206 (rid-based sticky): rid keys are not token sequences and are unaffected by token truncation.

## Test Plan
- Router (all three paths): follow-up pins to first-turn worker at threshold 0.5 with boundaries (repeat-selection hardened); the same follow-up provably lands elsewhere without boundaries; distinct conversations don't pin; boundary-first and no-boundary edge cases; empty-config identity
- Config: serde default/roundtrip; CLI list flow into `RouterConfig`
- `cargo +nightly fmt --all --check`; targeted module tests green (router 25, PD 12, gRPC selection 5, shared helper 3, config 38, CLI flow 1, pipeline 7); full clippy + suite in CI

<details><summary>Checklist</summary>

- [x] Format clean; lint self-audited (`#[expect]` with reason, no `unwrap` in prod paths)
- [x] Tests added and passing
- [x] Bindings updated (Python); Go unaffected
- [x] DCO sign-off, conventional commit
</details>

### CODE REVIEW

File: bindings/python/src/smg/router_args.py [hunk 1]
```diff
@@ -613,6 +614,19 @@ def add_cli_args(
                 " via --stream-request-bodies-over. 0 disables"
             ),
         )
+        routing_group.add_argument(
+            f"--{prefix}routing-token-boundaries",
+            type=int,
+            nargs="*",
+            action="extend",
+            default=[],
+            help=(
```
[the remark this answers is not in the document]
Comment (slin1237) [reply bindings/python/src/smg/router_args.py:642]: Fixed in a6031dd5: the prefixed variant now defaults to None (extend action still accumulates; argparse copies None to [] when the flag is present), so an unset --router-routing-token-boundaries falls back to the unprefixed backend value in from_cli_args, while an explicitly-empty prefixed flag still wins as a disable. Covered by test_routing_token_boundaries_fall_back_to_backend_value and test_routing_token_boundaries_prefixed_value_wins.

File: model_gateway/src/routers/grpc/common/stages/worker_selection.rs [hunk 2]
```diff
@@ -59,11 +62,28 @@ impl WorkerSelectionStage {
...
+    fn truncated_routing_tokens<'a>(&self, ids: &'a [u32]) -> Option<&'a [u32]> {
+        let ids = routing_tokens::truncate_slice(
+            ids,
+            &self.routing_token_boundaries,
+            metrics_labels::ROUTER_GRPC,
+        );
+        if ids.is_empty() {
+            None
+        } else {
+            Some(ids)
```
[the remark this answers is not in the document]
Comment (slin1237) [reply model_gateway/src/routers/grpc/common/stages/worker_selection.rs]: Aligned in a6031dd5: a boundary-emptied prefix now stays Some(&[]) on the gRPC path too (match rate 0 → min-load, no text fallback); only a genuinely token-free request maps to None, preserving the stage's pre-existing empty-token_ids convention. The boundary-first assertion in grpc_follow_up_pins_with_boundaries pins the new semantics.

File: model_gateway/src/routers/http/router.rs [hunk 3]
```diff
@@ -220,6 +224,18 @@ impl Router {
         }
     }
 
+    /// A boundary in first position yields an empty prefix: match rate 0,
+    /// min-load, and no full-text fallback into the string tree.
+    fn truncate_routing_tokens(&self, tokens: Option<Vec<u32>>) -> Option<Vec<u32>> {
```
[the remark this answers is not in the document]
Comment (slin1237) [reply model_gateway/src/routers/http/router.rs:229]: Discharged in 9261ca16 (branch rebased onto main with #2202): route_streaming_request now wraps its hint in truncate_routing_tokens, and a boundary-emptied hint falls back to buffering (Err with body unconsumed) — the stream was admitted on the hint's strength, so an empty shareable prefix must not route content-blind. Tests: boundary_emptied_hint_falls_back_to_buffered (body intact) and boundary_truncated_hint_streams_with_prefix_affinity (divergent tails past the boundary pin to one tenant; the untruncated ratio 32/65 sits below threshold, so only truncated selection explains the stick).

File: bindings/python/tests/test_arg_parser.py [hunk 4]
```diff
@@ -833,6 +833,26 @@ def test_prefixed_repeated_selector_accumulates(self):
...
+        assert router_args.routing_token_boundaries == [200091, 200038]
+
+    def test_routing_token_boundaries_prefixed_value_wins(self):
+        parser = argparse.ArgumentParser()
+        RouterArgs.add_cli_args(parser, use_router_prefix=True)
+        namespace = parser.parse_args(["--router-routing-token-boundaries", "200091", "200038"])
+
+        router_args = RouterArgs.from_cli_args(namespace, use_router_prefix=True)
+
+        assert router_args.routing_token_boundaries == [200091, 200038]
```
[the remark this answers is not in the document]
Comment (slin1237) [reply bindings/python/tests/test_arg_parser.py:866]: Strengthened in 9261ca16: the namespace now carries a competing unprefixed value ([111]) and the test asserts the explicit prefixed list wins — precedence is exercised, not vacuous.
